From fc6a9cf89d5e313a7a22d4ceccc207ab714aad6d Mon Sep 17 00:00:00 2001
From: Songjun Wu <songjun.wu@linux.intel.com>
Date: Tue, 26 Jan 2021 14:09:34 +0800
Subject: [PATCH] PONRTSYS-7128: i2c master status w 5

The previous change for interrupt is reverted.
Just pr_err() is changed to pr_debug().
Sometimes i2c HW will trigger dummy interrupt,
even the i2c transcation is finished, but it will
not cause any error.
---
 drivers/i2c/busses/i2c-lantiq.c | 30 ++++++++++++------------------
 drivers/i2c/busses/i2c-lantiq.h |  4 ----
 2 files changed, 12 insertions(+), 22 deletions(-)

diff --git a/drivers/i2c/busses/i2c-lantiq.c b/drivers/i2c/busses/i2c-lantiq.c
index 9953eb146404..44e73876ba5a 100644
--- a/drivers/i2c/busses/i2c-lantiq.c
+++ b/drivers/i2c/busses/i2c-lantiq.c
@@ -378,11 +378,11 @@ static void ltq_i2c_tx(struct ltq_i2c *priv)
 		}
 	}
 
-	if (priv->last_burst)
-		disable_burst_irq(priv);
-
 	/* write as word */
 	i2c_w32(word, txd);
+
+	if (priv->last_burst)
+		disable_burst_irq(priv);
 }
 
 /**
@@ -608,26 +608,20 @@ static irqreturn_t ltq_i2c_rx_thread_fn(int irq, void *dev_id)
 	mutex_unlock(&priv->rx_lock);
 	return IRQ_HANDLED;
 }
-
 static irqreturn_t ltq_i2c_isr_burst(int irq, void *dev_id)
 {
 	struct ltq_i2c *priv = dev_id;
 	struct i2c_msg *msg = priv->current_msg;
-	u32 i_raw;
-	int ret = IRQ_HANDLED;
 
-	i_raw = i2c_r32(mis);
+	int ret = IRQ_HANDLED;
 
-	if (i_raw & I2C_MIS_BREQ_INT) {
-		pr_debug("===B: flags[%x] status[%x] irq[%d]\n",
-			msg->flags, priv->status, irq);
-	} else if (i_raw & I2C_MIS_LBREQ_INT) {
-		priv->last_burst = LAST_BURST;
-		pr_debug("===LB: flags[%x] status[%x] irq[%d]\n",
-			msg->flags, priv->status, irq);
-	} else {
-		return ret;
-	}
+	priv->last_burst = (irq == priv->irq_lb) ? LAST_BURST : BURST;
+	if (priv->last_burst)
+		pr_debug("===LB: flags[%x] status[%x]\n",
+			msg->flags, priv->status);
+	else
+		pr_debug("===B: flags[%x] status[%x]\n",
+			msg->flags, priv->status);
 
 	if (msg->flags & I2C_M_RD) {
 		switch (priv->status) {
@@ -645,7 +639,7 @@ static irqreturn_t ltq_i2c_isr_burst(int irq, void *dev_id)
 			break;
 		default:
 			disable_burst_irq(priv);
-			pr_err("===Status R %d\n", priv->status);
+			pr_debug("===Status R %d\n", priv->status);
 			break;
 		}
 	} else {
diff --git a/drivers/i2c/busses/i2c-lantiq.h b/drivers/i2c/busses/i2c-lantiq.h
index 02b9a449ffca..7f84dcdd56a3 100644
--- a/drivers/i2c/busses/i2c-lantiq.h
+++ b/drivers/i2c/busses/i2c-lantiq.h
@@ -247,9 +247,5 @@ struct lantiq_reg_i2c {
 /* TX_END */
 #define I2C_P_IRQSS_TX_END 0x00000020
 
-/* Last Burst */
-#define I2C_MIS_LBREQ_INT	BIT(2)
-/* Burst */
-#define I2C_MIS_BREQ_INT	BIT(3)
 
 #endif /* I2C_LANTIQ_H */
